m2cgen/interpreters/interpreter.py

```diff
@@ -1,18 +1,64 @@
 import re
 
 from m2cgen import ast
+from m2cgen.interpreters.code_generator import BaseCodeGenerator
 
 
-class BaseInterpreter:
-
-    def __init__(self, cg, feature_array_name="input"):
-        self._cg = cg
-        self._feature_array_name = feature_array_name
+class BaseAstInterpreter:
 
     def interpret(self, expr):
         return self._do_interpret(expr)
 
-    # Default method implementations
+    # Private methods implementing visitor pattern
+
+    def _pre_interpret_hook(self, expr, **kwargs):
+        return expr, kwargs
+
+    def _do_interpret(self, expr, **kwargs):
+        # Hook which allows to override kwargs and to return custom result.
+        result, kwargs = self._pre_interpret_hook(expr, **kwargs)
+
+        # If returned result is the same as incoming expr - it means that we
+        # still need to process it.
+        if result is not expr:
+            return result
+
+        try:
+            handler = self._select_handler(expr)
+        except NotImplementedError:
+            if isinstance(expr, ast.TransparentExpr):
+                return self._do_interpret(expr.expr,  **kwargs)
+            raise
+        return handler(expr, **kwargs)
+
+    def _select_handler(self, expr):
+        handler_name = self._handler_name(type(expr))
+        if hasattr(self, handler_name):
+            return getattr(self, handler_name)
+        raise NotImplementedError(
+            "No handler found for {}".format(type(expr).__name__))
+
+    @staticmethod
+    def _handler_name(expr_tpe):
+        expr_name = BaseAstInterpreter._normalize_expr_name(expr_tpe.__name__)
+        return "interpret_" + expr_name
+
+    @staticmethod
+    def _normalize_expr_name(name):
+        return re.sub("(?!^)([A-Z]+)", r"_\1", name).lower()
+
+
+class BaseAstToCodeInterpreter(BaseAstInterpreter):
```

Comment:
Why do we even need this intermediate entity?

Reply:
Otherwise mixins have attribute `_cg` which they are not aware of what it is.